api-authentication/server/server.js

```diff
@@ -0,0 +1,80 @@
+import 'dotenv/config';
+import pg from 'pg';
+import argon2 from 'argon2';
+import express from 'express';
+import jwt from 'jsonwebtoken';
+import { ClientError, errorMiddleware } from './lib/index.js';
+
+const db = new pg.Pool({
+  connectionString: process.env.DATABASE_URL,
+  ssl: {
+    rejectUnauthorized: false
+  }
+});
+
+const app = express();
+app.use(express.json());
+
+app.post('/api/auth/sign-up', async (req, res, next) => {
+  try {
+    const { username, password } = req.body;
+    if (!username || !password) {
+      throw new ClientError(400, 'username and password are required fields');
+    }
+    const hashedPassword = await argon2.hash(password);
+    const sql = `
+      insert into "users" ("username", "hashedPassword")
+      values ($1, $2)
+      returning *
+    `;
+    const params = [username, hashedPassword];
+    const result = await db.query(sql, params);
+    const [user] = result.rows;
+    res.status(201).json(user);
+  } catch (err) {
+    next(err);
+  }
+});
+
+/**
+ * Checks reuqest for username and password and vierfies with
+ * argon2 against the database and returns a token, and object with userId and username.
+ */
+app.post('/api/auth/sign-in', async (req, res, next) => {
+  try {
+    const { username, password } = req.body;
+    if (!username || !password) {
+      throw new ClientError(401, 'invalid login');
+    }
+    const sql = `
+                select "userId", "hashedPassword"
+                from "users"
+                where "username" = $1
+                `;
+    const data = await db.query(sql, [username]);
+    if (!data.rows[0]) {
+      throw new ClientError(401, 'username does not exits');
```

Comment:
For security reasons, we usually don't tell the user whether the username or password is the part that is wrong. Instead, we just always say "Invalid login."